Buildpack support for sBOM (#269)
buildpacks/spec: buildpack.md

    - [Purpose](#purpose-2)
    - [Process](#process-2)
      - [Unmet Buildpack Plan Entries](#unmet-buildpack-plan-entries)

>       - [Software-Bill-of-Materials](#software-bill-of-materials)

Reviewer: ```suggestion
      - [Software Bill of Materials](#software-bill-of-materials)
```

| `<layers>/build.toml`                    | Build metadata (see [build.toml](#buildtoml-toml))
| `<layers>/store.toml`                    | Persistent metadata (see [store.toml](#storetoml-toml))
| `<layers>/<layer>.toml`                  | Layer metadata (see [Layer Content Metadata](#layer-content-metadata-toml))

> | `<layers>/<layer>.bom.<ext>`             | Layer standardized Bill of Materials (see [Bills-of-Materials](#bill-of-materials))

Author: Please see 41d7cfd

The following table illustrates the behavior depending on the value of each flag.
Note that the lifecycle only restores layers from the cache, never from the previous image.

`build`   | `cache`  | `launch` | Metadata and SBOM** Restored | Layer Restored
----------|----------|----------|------------------------------|---------------------
true      | true     | true     | Yes - from the app image     | Yes* - from the cache
true      | true     | false    | Yes - from the cache         | Yes - from the cache
true      | false    | true     | No                           | No
true      | false    | false    | No                           | No
false     | true     | true     | Yes - from the app image     | Yes* - from the cache
false     | true     | false    | Yes - from the cache         | Yes - from the cache
false     | false    | true     | Yes - from the app image     | No
false     | false    | false    | No                           | No

\* The metadata and layer are restored only if the layer SHA recorded in the previous image matches the layer SHA recorded in the cache.

> \** Only SBOM files associated with a layer are restored. Launch-level and build-level SBOM files must be re-created on each build.

Reviewer 2: nit: Can this sentence be broken into a new line (when rendered)? It is currently difficult to find where ** annotation information is declared.

Author: Added in cc19f13

Before a given re-build:
- If a launch layer is marked `cache = false` and `build = false` in the previous image metadata, the lifecycle:
  - MUST restore Layer Content Metadata to `<layers>/<layer>.toml`, excluding the `[types]` table.

>   - MUST restore any provided standardized Bill of Materials to `<layers>/<layer>.bom.<ext>`.

Reviewer: We should note that we are only restoring BOM associated with layers instead of any BOM files.

Author: Please see 41d7cfd

1. Compile the application source code into object code.
1. Remove application source code that is not necessary for launch.
1. Provide start command in `<layers>/launch.toml`.
1. Write a partial Software Bill of Materials to `<layers>/<layer>.sbom.<ext>` describing any dependencies provided in the layer.
1. Write a partial Software Bill of Materials to `<layers>/launch.sbom.<ext>` describing any provided application dependencies not associated with a layer.

> 1. Write a partial Software Bill of Materials to `<layers>/build.sbom.<ext>` describing any provided build dependencies not associated with a layer.

Reviewer: I think one thing that I know people have gotten confused about, is where they should write sbom for libraries that were included in an output compiled binary. For eg go libraries in a go binary. Those should technically be in the launch sbom as they are present in the final app image. On the other hand tools like a package manager might be better off in the build sbom. Not sure if we should clarify here or in the docs.

#### Bills-of-Materials

> Buildpacks MAY write standardized Bill of Materials (sBOM) files with extension `<ext>`, where `<ext>` MUST denote an sBOM media type based on Internet Assigned Numbers Authority (IANA) [assigned media types](https://www.iana.org/assignments/media-types/media-types.xhtml). The currently supported media types and their expected file extensions are as follows:

Reviewer: It might be confusing to use the s in sbom to mean `standardized` instead of `software`.

Reviewer: FWIW CycloneDx does make a distinction as it provides a bill of materials for non software packages as well like services and hardware.

Author: I'll change sBOM -> BOM everywhere.

Author: Please see dac2a68

Author: Following discussion in Core Team Sync, `a06a415` uses `SBOM` everywhere instead.

 | sBOM Media Type                  | File Extension
 |----------------------------------|----------------------------------------------
 | `application/vnd.cyclonedx+json` | `cdx.json`
 | `application/spdx+json`          | `spdx.json`
 | `application/vnd.syft+json`      | `syft.json`

When the build is complete, an sBOM describing the app image MAY be generated for auditing purposes.
If generated, this BOM MUST contain all `<layer>.bom.<ext>` files for each `launch = true` layer at the end of each `/bin/build` execution, as well as `launch.bom.<ext>` if provided, in adherence with the process and data format outlined in the [Platform Interface Specification](platform.md).

> When the build is complete, a build sBOM describing the build container MAY be generated for auditing purposes.

Reviewer 2: This sentence and that in line 562 appear to be duplicates until you realize that one is talking about launch vs build. Maybe we can emphasize the difference through style:

```suggestion
When the build is complete, a **build sBOM** describing the build container MAY be generated for auditing purposes.
```

> If generated, this BOM MUST contain all `<layer>.bom.<ext>` files for each `launch = false` layer at the end of each `/bin/build` execution, as well as `build.bom.<ext>` if provided, in adherence with the process and data format outlined in the [Platform Interface Specification](platform.md).

Reviewer 2: 1. Same comment as [this](https://github.com/buildpacks/spec/pull/268/files#r746061076) about leakiness.
2. Where do we talk about _data format_ in the platform spec? Do we mean file structure?

Author: @jromero for (1) above the link did not work. Are you talking about MUST vs MAY?

For (2), I'll change "data format" to "file structure". The old wording is a holdover from the old unstructured bom.

Author: Please see e79d4a0

Author: Ah re: (1) I see it here:

> Probably future refactor discussion here but...

> This is an interesting current leaky abstraction between the platform and buildpack APIs. As a platform implementor, I don't think I care about the contents of `<layers>/<buidpack-id>`. I wonder if this is the right spec to define this...

I agree, it's a bit weird. I guess it's helpful for the buildpack to know that the sBOM it writes is going "somewhere". We should bring this up as part of a larger spec refactor discussion. I would vote not to attempt a refactor as part of this PR.

#### Bills-of-Materials

Buildpacks MAY write standardized Bill of Materials (sBOM) files with extension `<ext>`, where `<ext>` MUST denote an sBOM media type based on Internet Assigned Numbers Authority (IANA) [assigned media types](https://www.iana.org/assignments/media-types/media-types.xhtml). The currently supported media types and their expected file extensions are as follows:

 | sBOM Media Type                  | File Extension
 |----------------------------------|----------------------------------------------
 | `application/vnd.cyclonedx+json` | `cdx.json`
 | `application/spdx+json`          | `spdx.json`

>  | `application/vnd.syft+json`      | `syft.json`

Author: Added Syft based on https://github.com/buildpacks/lifecycle/issues/755

> #### Bills-of-Materials

Reviewer: ```suggestion
#### Bill-of-Materials
```

Author: Please see 313fec2

Author: Please see a06a415

clear-env = false
description = "<buildpack description>"
keywords = [ "<string>" ]

> sbom = [ "<string>" ]

Reviewer 2: Minor: At first glance the key was a little too generic that I thought I could somehow add additional sbom data via this key when in reality it's more about defining supported media types.

I would propose changing it to something like `sbom-types`.

Author: @samj1912 WDYT?

Reviewer: `sbom-formats` maybe?

Author: Please see 12aff6f

clear-env = false
description = "<buildpack description>"
keywords = [ "<string>" ]

> sbom-formats = [ "<string>" ]

Reviewer: Given that we now use `bom` everywhere else, should we also rename this to `bom-formats`?

Author: Please see 313fec2

Author: Please see a06a415

- `type` - This MAY use the SPDX 2.1 license expression, but is not limited to identifiers in the SPDX Licenses List.
- `uri` - If this buildpack is using a nonstandard license, then this key MAY be specified in lieu of or in addition to `type` to point to the license.

**The buildpack sBOM:**

*Key: `sbom = [ "<string>" ]`*

>  - MUST be a supported sBOM media type as described in [Bills-of-Materials](#bills-of-materials).

Reviewer 2: Given that this is a list, we should be referring to them as plural no?

```suggestion
 - MUST be supported sBOM media types as described in [Bills-of-Materials](#bills-of-materials).
```
